grep, tail, head: add option to keep header

bernt-matthias opened the pull request:
## What did you do? 

1. Add an option to the head, tail, and grep tools that allows to keep the 1st line (the header).
2. Remove python / perl wrappers

## Why did you make this change?

1. https://github.com/galaxyproject/tools-iuc/issues/3511 https://github.com/galaxyproject/galaxy/issues/5149
2. The wrappers just called the command lines and are therefore not necessary


## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

## For UI Components
- [ ] I've included a screenshot of the changes

bernt-matthias committed:
grep: add option to keep header

https://github.com/galaxyproject/tools-iuc/issues/3511

bernt-matthias committed:
consider header in tail/head wrapper

and remove pl wrapper

mvdbeek commented on tools/filters/headWrapper.xml:
```diff
@@ -4,11 +4,21 @@
   </edam_operations>
   <command>
-perl '$__tool_directory__/headWrapper.pl' '$input' $lineNum '$out_file1'
+    set -eo pipefail;
```
I guess we need to add a profile version here so that the non-zero exit code fails the job ?
We also need to bump the version, add the tool versions to https://github.com/mvdbeek/galaxy/blob/926268ce6e1acd3db4e51ceb72e689899c545b9c/lib/galaxy/tools/__init__.py#L195 and wrap the command block in CDATA

bernt-matthias replied:
No problem. Should this be done as well for `Grep1`? What is the rule whether it should be included?

mvdbeek replied:
If previously selected parameters don't change the output we should add it. If it does we should duplicate the tool in place.

bernt-matthias replied:
I think `Grep1` also fulfills these requirements. I already added it.

mvdbeek commented on tools/filters/grep.xml:
```diff
@@ -1,9 +1,9 @@
-<tool id="Grep1" name="Select" version="1.0.1">
+<tool id="Grep1" name="Select" version="1.0.2">
```
We also need to put this into https://github.com/mvdbeek/galaxy/blob/926268ce6e1acd3db4e51ceb72e689899c545b9c/lib/galaxy/tools/__init__.py#L195

mvdbeek commented on tools/filters/tailWrapper.xml:
```diff
@@ -1,10 +1,21 @@
   <description>lines from a dataset</description>
-  <command detect_errors="exit_code">
-perl '$__tool_directory__/tailWrapper.pl' '$input' $lineNum '$out_file1'</command>
+  <command detect_errors="exit_code"><![CDATA[
```
Same here, I think we need to bump the version and add the new version to https://github.com/mvdbeek/galaxy/blob/926268ce6e1acd3db4e51ceb72e689899c545b9c/lib/galaxy/tools/__init__.py#L195

bernt-matthias committed:
some updates for grep, tail, head

- CDATA for all
- exit code
- remove interpreter
- bump versions
- and add the tools to WORKFLOW_SAFE_TOOL_VERSION_UPDATES

bernt-matthias committed:
remove python wrapper for grep tool

and fix cli (`-n +2`) for head, tail, grep

mvdbeek commented on tools/filters/grep.xml:
```diff
@@ -1,36 +1,67 @@
-<tool id="Grep1" name="Select" version="1.0.1">
+<tool id="Grep1" name="Select" version="1.0.2">
   <description>lines that match an expression</description>
```
```suggestion
  <requirements>
    <requirement type="package" version="3.4">grep</requirement>
  </requirements>
  <description>lines that match an expression</description>
```

mvdbeek commented on tools/filters/tailWrapper.xml:
```diff
@@ -1,10 +1,21 @@
   <description>lines from a dataset</description>
-  <command detect_errors="exit_code">
-perl '$__tool_directory__/tailWrapper.pl' '$input' $lineNum '$out_file1'</command>
+  <command detect_errors="exit_code"><![CDATA[
```
```suggestion
  <requirements>
    <requirement type="package" version="8.31">coreutils</requirement>
  </requirements>
  <command detect_errors="exit_code"><![CDATA[
```

mvdbeek commented on tools/filters/headWrapper.xml:
```diff
@@ -1,23 +1,48 @@
-  <command>
-perl '$__tool_directory__/headWrapper.pl' '$input' $lineNum '$out_file1'
-  </command>
+  <command detect_errors="exit_code"><![CDATA[
```
```suggestion
  <requirements>
    <requirement type="package" version="8.31">coreutils</requirement>
  </requirements>
  <command detect_errors="exit_code"><![CDATA[
```

mvdbeek commented on tools/filters/grep.xml:
```diff
@@ -1,36 +1,67 @@
-<tool id="Grep1" name="Select" version="1.0.1">
+<tool id="Grep1" name="Select" version="1.0.2">
```
```suggestion
<tool id="Grep1" name="Select" version="1.0.2" profile="20.05">
```

mvdbeek commented on tools/filters/headWrapper.xml:
```diff
@@ -1,23 +1,48 @@
-<tool id="Show beginning1" name="Select first" version="1.0.0">
+<tool id="Show beginning1" name="Select first" version="1.0.1">
```
```suggestion
<tool id="Show beginning1" name="Select first" version="1.0.1" profile="20.05">
```

mvdbeek commented on tools/filters/tailWrapper.xml:
```diff
@@ -1,10 +1,21 @@
-<tool id="Show tail1" name="Select last" version="1.0.0">
+<tool id="Show tail1" name="Select last" version="1.0.1">
```
```suggestion
<tool id="Show tail1" name="Select last" version="1.0.1" profile="20.05">
```

mvdbeek reviewed (commented):
Hmm, if I run the tool test for Grep1 locally it fails (and we don't test all the built-in tools on PR currently). One test file is missing, and the other complains about unbalanced brackets:

# Test Results (powered by Planemo)

## Test Summary

<div class="progress">
  <div class="progress-bar progress-bar-success" style="width: 33.33333333333333%" aria-valuenow="1" aria-valuemin="0" aria-valuemax="3" data-toggle="tooltip" title="1 Passed">
  </div>
  <div class="progress-bar progress-bar-warning" style="width: 0.0%" aria-valuenow="0" aria-valuemin="0" aria-valuemax="3" data-toggle="tooltip" title="0 Skipped">
  </div>
  <div class="progress-bar progress-bar-danger" style="width: 66.66666666666666%" aria-valuenow="2" aria-valuemin="0" aria-valuemax="3" title="2 Failed or Errored">
  </div>
</div>

| Test State | Count |
| ---------- | ----- |
| Total      | 3 |
| Passed     | 1 |
| Error      | 0 |
| Failure    | 2 |
| Skipped    | 0 |


<details><summary>Failed Tests</summary>

* <details class="rcorners light-red"><summary class="light-red">&#10060; Grep1 (Test #1)</summary><div class="padded">

    **Problems**:
    * ```
      Output out_file1: Test output file (fs-grep.dat) is missing. If you are using planemo, try adding --update_test_data to generate it.
      ```
    **Command Line:**

    * ```console
      set -eo pipefail; cat '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/8/2/e/dataset_82e63542-3cf2-492c-ad72-90c26262ec74.dat' | grep -f '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/job_working_directory/000/2/configs/tmpuwq5t8rh'  > '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/1/1/4/dataset_1142ee44-a8c8-41cb-bc99-97af65e80e3c.dat'
      ```
    **Exit Code:**

    * ```console
      0
      ```
   **Job Parameters:**

   *   | Job parameter | Parameter value |
       | ------------- | --------------- |
       | \_\_input\_ext | ` "bed" ` |
       | chromInfo | ` "/Users/mvandenb/src/galaxy/tool-data/shared/ucsc/chrom/?.len" ` |
       | dbkey | ` "?" ` |
       | invert | ` "" ` |
       | keep\_header | ` "false" ` |
       | pattern | ` "^chr[0-9]*" ` |



* <details class="rcorners light-red"><summary class="light-red">&#10060; Grep1 (Test #3)</summary><div class="padded">

    **Problems**:
    * ```
      Job in error state.
      ```
    * ```
      Job in error state.
      ```
    **Command Line:**

    * ```console
      set -eo pipefail; head -n 1 '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/5/c/a/dataset_5ca23a52-d71f-49b5-8b3b-681596d75755.dat' > '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/6/e/a/dataset_6ea2dc8a-c002-4201-af0d-a5846ac7ad74.dat' && tail -n +2 '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/5/c/a/dataset_5ca23a52-d71f-49b5-8b3b-681596d75755.dat' | grep -f '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/job_working_directory/000/6/configs/tmpokbobuv1'  > '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/6/e/a/dataset_6ea2dc8a-c002-4201-af0d-a5846ac7ad74.dat'
      ```
    **Exit Code:**

    * ```console
      2
      ```
    **Standard Error:**

    * ```console
      grep: /private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/job_working_directory/000/6/configs/tmpokbobuv1:1: Unmatched [, [^, [:, [., or [=

      ```
   **Job Parameters:**

   *   | Job parameter | Parameter value |
       | ------------- | --------------- |
       | \_\_input\_ext | ` "bed" ` |
       | chromInfo | ` "/Users/mvandenb/src/galaxy/tool-data/shared/ucsc/chrom/?.len" ` |
       | dbkey | ` "?" ` |
       | invert | ` "" ` |
       | keep\_header | ` "true" ` |
       | pattern | ` "<>\\\"[]{}" ` |



</details>
<details><summary>Passed Tests</summary>

* <details class="rcorners light-green"><summary class="light-green">&#9989; Grep1 (Test #2)</summary><div class="padded">

    **Command Line:**

    * ```console
      set -eo pipefail; head -n 1 '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/8/d/8/dataset_8d8cac07-23e9-48c4-ac93-a4db03d21c07.dat' > '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/6/c/6/dataset_6c683416-0390-4ac7-84cd-8c772e140d6b.dat' && tail -n +2 '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/8/d/8/dataset_8d8cac07-23e9-48c4-ac93-a4db03d21c07.dat' | grep -f '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/job_working_directory/000/4/configs/tmp7rv7dgrl'  > '/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmp47d7v6s3/files/6/c/6/dataset_6c683416-0390-4ac7-84cd-8c772e140d6b.dat'
      ```
    **Exit Code:**

    * ```console
      0
      ```
   **Job Parameters:**

   *   | Job parameter | Parameter value |
       | ------------- | --------------- |
       | \_\_input\_ext | ` "bed" ` |
       | chromInfo | ` "/Users/mvandenb/src/galaxy/tool-data/shared/ucsc/chrom/?.len" ` |
       | dbkey | ` "?" ` |
       | invert | ` "" ` |
       | keep\_header | ` "true" ` |
       | pattern | ` "+" ` |



</details>

bernt-matthias committed:
Apply suggestions from code review

- add profile
- add requirement

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
fix test 3 and error detection

only error for exit code > 1:

bernt-matthias committed:
Merge branch 'topic/grep-header' of https://github.com/bernt-matthias/galaxy into topic/grep-header

mvdbeek reviewed (approved):
Very nice, all tests are passing now (locally)!
